### ISSUE TITLE
GA request for splunk-otel-android

### ISSUE THREAD

**Issue by jkwatson** [signalfx/gdi-specification#133]
**Which GDI repository do you wish to GA?**
[splunk-otel-android](https://github.com/signalfx/splunk-otel-android)

**Does the repository follow the latest tagged minor release in GDI specification?**

  - [x] Has an appropriate [maintainers team](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#teams). - https://github.com/orgs/signalfx/teams/gdi-java-maintainers
  - [x] [Permissions](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#permissions) set correctly.
  - [x] [Branch protection](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#branch-protection) in place.
  - [x] [Dependencies](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#dependencies) appropriately locked down : [link](https://github.com/signalfx/splunk-otel-android/blob/main/splunk-otel-android/build.gradle#L42-L64).
  - [x] [GitHub Applications](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#github-applications) set up per spec. 
  - [x] [Required Files](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#required-files) in place.
    - [x] [CHANGELOG.md](https://github.com/signalfx/splunk-otel-android/blob/main/CHANGELOG.md)
    - [x] [CODE_OF_CONDUCT.md](https://github.com/signalfx/splunk-otel-android/blob/main/CODE_OF_CONDUCT.md)
    - [x] [CONTRIBUTING.md](https://github.com/signalfx/splunk-otel-android/blob/main/CONTRIBUTING.md)
    - [x] [.github/CODEOWNERS](https://github.com/signalfx/splunk-otel-android/blob/main/.github/CODEOWNERS)
    - [x] [LICENSE](https://github.com/signalfx/splunk-otel-android/blob/main/LICENSE)
    - [x] [README.md](https://github.com/signalfx/splunk-otel-android/blob/main/README.md)
      - [x] Build status badge
      - [x] [Getting started](https://github.com/signalfx/splunk-otel-android#getting-started)
      - [x] [Troubleshooting](https://github.com/signalfx/splunk-otel-android#troubleshooting)
      - [x] [Link to official Splunk docs](https://github.com/signalfx/splunk-otel-android#splunk-android-rum-instrumentation)
      - [x] [License information](https://github.com/signalfx/splunk-otel-android#license-and-versioning)
    - [x] [SECURITY.md](https://github.com/signalfx/splunk-otel-android/blob/main/SECURITY.md)
  - [x] [Releases](../../specification/repository.md#github-releases) done to spec. [link](https://github.com/signalfx/splunk-otel-android/releases)
  - [x] [Real User Monitoring Library](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/configuration.md#real-user-monitoring-libraries)
    - [x] Documents all supported configuration parameters. [link](https://github.com/signalfx/splunk-otel-android#configuration) and [other link](https://github.com/signalfx/splunk-otel-android#additional-configbuilder-options)
    - [x] Documents how to configure manual instrumentation. [link to otel](https://github.com/signalfx/splunk-otel-android#apis-provided-by-the-splunkrum-instance)
    - [x] Documents supported instrumentation. [it's only okhttp](https://github.com/signalfx/splunk-otel-android#apis-provided-by-the-splunkrum-instance)
    - [x] Documents supported platforms. [link](https://github.com/signalfx/splunk-otel-android#prerequisites)

**How long has the GDI repository been public?**
Approximately 4 months

**Is the repository known to be used today?**
yes

**Is there a date by which this approval is needed?**
by the time we GA Mobile RUM

**Additional context**
none

**Comment by MrAlias** [comment]
> - [1] Application name, authentication token and either realm or the beacon URL MUST be provided by the user.
>   If any of these is missing, the RUM instrumentation library MUST fail to start.
> - [2] Implementations MUST enforce by default that `beaconEndpoint`s are https only, and reject/fail to start otherwise.
>   Implementations MAY offer an unsupported `allowInsecureBeacon` option (default false) that turns off that check.
> - [3] If both `realm` and `beaconEndpoint` are set, a warning saying that `realm` will be ignored SHOULD be logged.

- [x] Check implemented [here](https://github.com/signalfx/splunk-otel-android/blob/0b360388752e68ad4569e970fc2c0efdb2e371d1/splunk-otel-android/src/main/java/com/splunk/rum/Config.java#L155-L157)

**Comment by MrAlias** [comment]
> - [3] If both `realm` and `beaconEndpoint` are set, a warning saying that `realm` will be ignored SHOULD be logged.

This isn't a requirement, but it also does not look implemented. @jkwatson can you validate if I missed something here?

**Comment by MrAlias** [comment]
> - RUM library MUST use the Zipkin v2 JSON span exporter by default

- [x] This looks to be the case: https://github.com/signalfx/splunk-otel-android/blob/141d16fbc29f6b0eb68b18679181fedb82bc9c6f/splunk-otel-android/src/main/java/com/splunk/rum/RumInitializer.java#L221

**Comment by MrAlias** [comment]
> - RUM library MUST limit the number of sent spans to 100 in a 30 second window per `component` attribute value

- [x] Throttling tested [here](https://github.com/signalfx/splunk-otel-android/blob/68d29497f638d6c96cb4620e7be3c08d3b57204b/splunk-otel-android/src/test/java/com/splunk/rum/ThrottlingExporterTest.java#L71), with defaults [here](https://github.com/signalfx/splunk-otel-android/blob/68d29497f638d6c96cb4620e7be3c08d3b57204b/splunk-otel-android/src/main/java/com/splunk/rum/ThrottlingExporter.java#L117-L118)

**Comment by jkwatson** [comment]
> > * [2] Implementations MUST enforce by default that `beaconEndpoint`s are https only, and reject/fail to start otherwise.
> >   Implementations MAY offer an unsupported `allowInsecureBeacon` option (default false) that turns off that check.
> 
> ❌ I'm not seeing where this is enforced. @jkwatson can you confirm it is implemented?

We don't implement this on Android because it's built in to the way that Android itself works. It's not something that can be done at the library level, but is only done at the application level, which we obviously cannot control.

**Comment by MrAlias** [comment]
> > > * [2] Implementations MUST enforce by default that `beaconEndpoint`s are https only, and reject/fail to start otherwise.
> > >   Implementations MAY offer an unsupported `allowInsecureBeacon` option (default false) that turns off that check.
> > 
> > 
> > x I'm not seeing where this is enforced. @jkwatson can you confirm it is implemented?
> 
> We don't implement this on Android because it's built in to the way that Android itself works. It's not something that can be done at the library level, but is only done at the application level, which we obviously cannot control.

Does this mean the specification needs to be updated to say this is a requirement if the system allows this kind of configuration enforcement?

**Comment by jkwatson** [comment]
> > * [3] If both `realm` and `beaconEndpoint` are set, a warning saying that `realm` will be ignored SHOULD be logged.
> 
> ❗ This isn't a requirement, but it also does not look implemented. @jkwatson can you validate if I missed something here?

We haven't implemented this, no. It would be simple to add this warning, though, if you think this is a blocker.

**Comment by MrAlias** [comment]
> > > * [3] If both `realm` and `beaconEndpoint` are set, a warning saying that `realm` will be ignored SHOULD be logged.
> > 
> > 
> > exclamation This isn't a requirement, but it also does not look implemented. @jkwatson can you validate if I missed something here?
> 
> We haven't implemented this, no. It would be simple to add this warning, though, if you think this is a blocker.

Not a blocker, but probably a good things to add given it can be added.

**Comment by jkwatson** [comment]
> > > > * [3] If both `realm` and `beaconEndpoint` are set, a warning saying that `realm` will be ignored SHOULD be logged.
> > > 
> > > 
> > > exclamation This isn't a requirement, but it also does not look implemented. @jkwatson can you validate if I missed something here?
> > 
> > 
> > We haven't implemented this, no. It would be simple to add this warning, though, if you think this is a blocker.
> 
> Not a blocker, but probably a good things to add given it can be added.

@mateuszrzeszutek can you take this as an action item?

**Comment by MrAlias** [comment]
@flands can you verify the [permissions](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#permissions) and [branch protections](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#branch-protection) please?

**Comment by MrAlias** [comment]
> SHOULD have Codecov GitHub App configured

I don't see this configured. If I recall correctly, this was not possible for Android, right? <- @jkwatson 

> SHOULD have Lychee Link Checker GitHub Action configured

Can we add this?

**Comment by jkwatson** [comment]
> > MUST have Jira GitHub App configured
> 
> I'm not sure where this is configured. @jkwatson can you verify this was configured?

I also don't know, but @flands told me it was done.

**Comment by MrAlias** [comment]
> > > MUST have Jira GitHub App configured
> > 
> > 
> > I'm not sure where this is configured. @jkwatson can you verify this was configured?
> 
> I also don't know, but @flands told me it was done.

Sounds good :+1:

**Comment by MrAlias** [comment]
> MUST update all examples in the Splunk OpenTelemetry example repository that depends on the repository to use the latest release.

- [x] no examples to update.

**Comment by MrAlias** [comment]
> All Splunk distributions of OpenTelemetry,
> 
> - MUST set the following OpenTelemetry resource attributes according to the [OpenTelemetry Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#telemetry-sdk):
>   - `telemetry.sdk.name`
>   - `telemetry.sdk.version`
>   - `telemetry.sdk.language`
> 
> - SHOULD set the following resource attributes when applicable:
>   - `telemetry.auto.version`

@jkwatson is this done natively in the Java SDK?

**Comment by MrAlias** [comment]
> > All Splunk distributions of OpenTelemetry,
> > 
> > * MUST set the following OpenTelemetry resource attributes according to the [OpenTelemetry Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#telemetry-sdk):
> >   
> >   * `telemetry.sdk.name`
> >   * `telemetry.sdk.version`
> >   * `telemetry.sdk.language`
> > * SHOULD set the following resource attributes when applicable:
> >   
> >   * `telemetry.auto.version`
> 
> @jkwatson is this done natively in the Java SDK?

> Note: this section does not apply to Real User Monitoring libraries, as they do not use the OpenTelemetry Resource.

:facepalm: NVM

**Comment by jkwatson** [comment]
> > SHOULD have Codecov GitHub App configured
> 
> I don't see this configured. If I recall correctly, this was not possible for Android, right? <- @jkwatson

I'm not sure if it's possible or not...but I personally have philosophical problems with using it. I'll leave this one to @mateuszrzeszutek and the team to decide if they want to use this as a proxy for test quality.

> 
> > SHOULD have Lychee Link Checker GitHub Action configured
> 
> Can we add this?

I don't no of a reason why not! @mateuszrzeszutek another action item for y'all.

**Comment by MrAlias** [comment]
> Real User Monitoring (RUM) libraries MUST set the `service.name` resource attribute to the value of the `applicationName`
> configuration property. This is the only resource attribute that RUM libraries are supposed to set because it's the only one
> the Zipkin exporter can understand.

- [x] https://github.com/signalfx/splunk-otel-android/blob/141d16fbc29f6b0eb68b18679181fedb82bc9c6f/splunk-otel-android/src/main/java/com/splunk/rum/RumInitializer.java#L192

**Comment by MrAlias** [comment]
> The following attributes MUST be added to all spans produced by RUM libraries:
> 
> | Name                  | Type   | Description                                                                                                       |
> | ----                  | ----   | -----------                                                                                                       |
> | `app`                 | string | The value of the `applicationName` property configured in `SplunkRum`; same as `service.name` resource attribute. |
> | `splunk.rumSessionId` | string | The RUM session ID                                                                                                |
> | `splunk.rum.version`  | string | Version of the RUM library                                                                                        |
> | `component`           | string | Name of the instrumentation that produced this span                                                               |

- [x] `app` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L62
- [x] `splunk.rumSessionId` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L63
- [x] `splunk.rum.version` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L64
-  [ ] `component` I found the key defined [here](https://github.com/signalfx/splunk-otel-android/blob/77a81c8f7b5969708e093da6686cacdfe33d3659/splunk-otel-android/src/main/java/com/splunk/rum/SplunkRum.java#L54), but do not see where it is added to a span outside of an exception. @jkwatson can you confirm this is/isn't added to all spans?

**Comment by jkwatson** [comment]
> > The following attributes MUST be added to all spans produced by RUM libraries:
> > 
> > 
> > 
> > Name
> > Type
> > Description
> > 
> > 
> > 
> > 
> > `app`
> > string
> > The value of the `applicationName` property configured in `SplunkRum`; same as `service.name` resource attribute.
> > 
> > 
> > `splunk.rumSessionId`
> > string
> > The RUM session ID
> > 
> > 
> > `splunk.rum.version`
> > string
> > Version of the RUM library
> > 
> > 
> > `component`
> > string
> > Name of the instrumentation that produced this span
> 
> * [x]  `app` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L62
> * [x]  `splunk.rumSessionId` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L63
> * [x]  `splunk.rum.version` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L64
> * [ ]  `component` I found the key defined [here](https://github.com/signalfx/splunk-otel-android/blob/77a81c8f7b5969708e093da6686cacdfe33d3659/splunk-otel-android/src/main/java/com/splunk/rum/SplunkRum.java#L54), but do not see where it is added to a span outside of an exception. @jkwatson can you confirm this is/isn't added to all spans?

https://github.com/signalfx/splunk-otel-android/search?q=COMPONENT_KEY

**Comment by MrAlias** [comment]
> Mobile RUM libraries (iOS, Android) MUST add device and system information to all spans:
> 
> | Name                | Type   | Description        |
> | ----                | ----   | -----------        |
> | `device.model.name` | string | Name of the device |
> | `os.type`           | string | `iOS` or `Android` |
> | `os.version`        | string | Version of the OS  |

- [x] `device.model.name`  https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L66
- [x] `os.type` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L69
- [x] `os.version` https://github.com/signalfx/splunk-otel-android/blob/02c8e16ecef3845294fece799e521c651f7b7cbc/splunk-otel-android/src/main/java/com/splunk/rum/RumAttributeAppender.java#L70

**Comment by MrAlias** [comment]
> It's not quite added to _all_ spans (for instance, we don't set it on user-provided spans/events), but all the ones that it's supposed to be set on, it is set on. ;)

SGTM :+1:

**Comment by MrAlias** [comment]
Remaining tasks:

- [ ] @flands to check permissions and branch protections.
- [ ] https://github.com/signalfx/gdi-specification/pull/146

Outstanding "nice to haves"

- [ ] https://github.com/signalfx/splunk-otel-android/issues/181
- [ ] https://github.com/signalfx/splunk-otel-android/issues/182

**Comment by jkwatson** [comment]
#181 has been completed (looks like I can't edit your checklist, @MrAlias )

**Comment by MrAlias** [comment]
The RUM configuration is planned to be set to a "Feature Freeze" version, but not release as stable. Because of this we will sign off on the splunk-otel-android repository publishing compliance with the GDI specification for the upcoming version, but it is not going to itself be released as stable given the instability in the configuration specification.

The last things we need to accomplish before we can close this sign off that the repository meets the permissions (@flands) and a new release of the gdi-specification needs to be made.

**Comment by flands** [comment]
> @flands can you verify the [permissions](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#permissions) and [branch protections](https://github.com/signalfx/gdi-specification/blob/v1.1.0/specification/repository.md#branch-protection) please?

Done!

> MUST have Jira GitHub App configured

Done!

**Comment by MrAlias** [comment]
Now that [Release v1.2.0](https://github.com/signalfx/gdi-specification/releases/tag/v1.2.0) is out this issue can be closed. The splunk-otel-android repository is authorized to say they are compliant with v1.2.0 of the GDI specification and add a badge to their README saying so (i.e. https://img.shields.io/badge/GDI-1.2.0-blueviolet?style=for-the-badge). However, since the RUM configuration is not yet released as stable (it is waiting on upstream OTel to stabilize), the splunk-otel-android repository is not authorized to make a 1.0.0 stable release.

When the RUM configuration becomes stable a new issue can be created requesting a GA release.
